File: compiler/sources_registry.py
"""Registry of external source folders symlinked into data/raw/.

Lets a user register additional folders that live elsewhere on disk (an
email export, a synced Drive folder, a notes vault) without moving or
copying anything. Each registered folder becomes a symlink under
RAW_DIR/<slug>/ pointing at the real path -- discover_raw_source_files()
already walks RAW_DIR recursively (synthesizer.py), so a registered folder
is picked up by the existing pipeline with no changes to extraction,
synthesis, or linking. This module only manages the registry file and the
symlinks; it never touches pipeline logic.
"""

from __future__ import annotations

import json
import re
import time
import uuid
from pathlib import Path
from typing import Any

from models import PROJECT_ROOT, RAW_DIR

SOURCES_FILE = PROJECT_ROOT / "data" / "sources.json"


def _slugify_label(label: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", label.strip().lower()).strip("-")
    return slug or "source"


def _unique_link_name(label: str, existing_names: set[str]) -> str:
    base = _slugify_label(label)
    name = base
    counter = 2
    while name in existing_names:
        name = f"{base}-{counter}"
        counter += 1
    return name


def _load() -> dict[str, Any]:
    if not SOURCES_FILE.is_file():
        return {"version": 1, "sources": []}
    try:
        data = json.loads(SOURCES_FILE.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return {"version": 1, "sources": []}
    data.setdefault("sources", [])
    return data


def _save(data: dict[str, Any]) -> None:
    SOURCES_FILE.parent.mkdir(parents=True, exist_ok=True)
    SOURCES_FILE.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def _count_files(path: Path) -> int:
    if not path.is_dir():
        return 0
    try:
        return sum(1 for p in path.rglob("*") if p.is_file() and not p.name.startswith("."))
    except OSError:
        return 0


def list_sources() -> list[dict[str, Any]]:
    """Registered source folders, enriched with live existence/file-count."""
    sources = _load()["sources"]
    enriched = []
    for entry in sources:
        target = Path(entry["path"])
        enriched.append(
            {
                **entry,
                "exists": target.is_dir(),
                "file_count": _count_files(target),
            }
        )
    return enriched


def add_source(path: str, label: str | None = None) -> dict[str, Any]:
    if not path or not path.strip():
        raise ValueError("Path is required")

    resolved = Path(path).expanduser().resolve()
    if not resolved.is_dir():
        raise ValueError(f"Not a directory: {resolved}")

    raw_resolved = RAW_DIR.resolve()
    if resolved == raw_resolved or raw_resolved in resolved.parents:
        raise ValueError("Cannot add data/raw/ (or a folder inside it) as a source")
    if resolved in raw_resolved.parents:
        raise ValueError("Cannot add a folder that contains data/raw/ as a source")

    data = _load()
    sources = data["sources"]
    if any(Path(entry["path"]).resolve() == resolved for entry in sources):
        raise ValueError("This folder is already registered")

    existing_names = {entry["link_name"] for entry in sources}
    link_name = _unique_link_name(label or resolved.name, existing_names)

    entry = {
        "id": uuid.uuid4().hex[:12],
        "label": (label or resolved.name).strip() or resolved.name,
        "path": str(resolved),
        "link_name": link_name,
        "enabled": True,
        "added_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }
    sources.append(entry)
    _save(data)
    sync_symlinks()
    return entry


def remove_source(source_id: str) -> bool:
    data = _load()
    sources = data["sources"]
    removed_entry = next((entry for entry in sources if entry["id"] == source_id), None)
    if removed_entry is None:
        return False
    data["sources"] = [entry for entry in sources if entry["id"] != source_id]
    _save(data)
    _remove_mirror(RAW_DIR / removed_entry["link_name"])
    sync_symlinks()
    return True


def set_enabled(source_id: str, enabled: bool) -> dict[str, Any] | None:
    data = _load()
    for entry in data["sources"]:
        if entry["id"] == source_id:
            entry["enabled"] = enabled
            _save(data)
            sync_symlinks()
            return entry
    return None


def _remove_mirror(root: Path) -> None:
    """Remove only the symlinks (and now-empty directories) this module
    created under `root` -- never a real file, and never a directory that
    still has something else in it."""
    if root.is_symlink():
        root.unlink()
        return
    if not root.is_dir():
        return
    for path in sorted(root.rglob("*"), key=lambda p: len(p.parts), reverse=True):
        if path.is_symlink():
            path.unlink()
        elif path.is_dir():
            try:
                path.rmdir()  # no-op (raises, caught) if anything real remains
            except OSError:
                pass
    try:
        root.rmdir()
    except OSError:
        pass


def _mirror_source(link_root: Path, target: Path) -> None:
    """Recreate link_root as a directory of per-file symlinks mirroring
    target's tree.

    A single symlink to an external directory is invisible to
    discover_raw_source_files()'s Path.rglob() scan, which does not descend
    into symlinked directories -- so each file gets its own symlink instead,
    inside a real (or already-existing) directory at link_root. Re-running
    this brings link_root back in sync with target: new files in target get
    a new symlink, files removed from target lose their mirrored symlink.
    """
    link_root.mkdir(parents=True, exist_ok=True)
    live_rel_paths: set[Path] = set()

    for path in target.rglob("*"):
        if not path.is_file():
            continue
        rel = path.relative_to(target)
        live_rel_paths.add(rel)
        dest = link_root / rel
        dest.parent.mkdir(parents=True, exist_ok=True)
        if dest.is_symlink():
            try:
                if dest.resolve() == path.resolve():
                    continue
            except OSError:
                pass
            dest.unlink()
        elif dest.exists():
            continue  # a real file already occupies this relative path
        try:
            dest.symlink_to(path)
        except OSError:
            continue

    for path in sorted(link_root.rglob("*"), key=lambda p: len(p.parts), reverse=True):
        rel = path.relative_to(link_root)
        if path.is_symlink():
            if rel not in live_rel_paths:
                path.unlink()
        elif path.is_dir():
            try:
                path.rmdir()
            except OSError:
                pass


def sync_symlinks() -> None:
    """Make RAW_DIR's managed source mirrors match the registry exactly.

    Only ever creates/removes symlinks (and the directories that hold them)
    under names tracked in the registry -- real files and directories a
    user placed in data/raw/ directly are never touched.
    """
    RAW_DIR.mkdir(parents=True, exist_ok=True)
    data = _load()
    sources = data["sources"]
    all_names = {entry["link_name"] for entry in sources}
    wanted = {
        entry["link_name"]: Path(entry["path"])
        for entry in sources
        if entry.get("enabled", True)
    }

    for child in RAW_DIR.iterdir():
        if child.name in all_names and child.name not in wanted:
            _remove_mirror(child)

    for name, target in wanted.items():
        link_root = RAW_DIR / name
        if link_root.is_symlink():
            link_root.unlink()  # replace a top-level symlink (e.g. from an older layout) with a mirror directory
        elif link_root.exists() and not link_root.is_dir():
            continue  # a real file already occupies this name -- don't clobber it
        if not target.is_dir():
            continue
        _mirror_source(link_root, target)

File: backend/src/lib/sourcesRegistry.ts
/**
 * Port of compiler/sources_registry.py -- registers external folders as a
 * mirrored tree of per-file symlinks under data/raw/, so the Python
 * compiler's discover_raw_source_files() (which walks RAW_DIR recursively,
 * same non-symlink-following-directory caveat that motivated per-file
 * mirroring in the Python original) picks every file up unchanged.
 */
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { RAW_DIR, SOURCES_FILE } from '../paths';
import { walkEntries } from './fsWalk';

export interface SourceEntry {
  id: string;
  label: string;
  path: string;
  link_name: string;
  enabled: boolean;
  added_at: string;
}

export interface SourceEntryWithStatus extends SourceEntry {
  exists: boolean;
  file_count: number;
}

interface Registry {
  version: number;
  sources: SourceEntry[];
}

export class SourceError extends Error {}

function slugifyLabel(label: string): string {
  const slug = label
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
  return slug || 'source';
}

function uniqueLinkName(label: string, existing: Set<string>): string {
  const base = slugifyLabel(label);
  let name = base;
  let counter = 2;
  while (existing.has(name)) {
    name = `${base}-${counter}`;
    counter += 1;
  }
  return name;
}

function load(): Registry {
  if (!fs.existsSync(SOURCES_FILE)) {
    return { version: 1, sources: [] };
  }
  try {
    const data = JSON.parse(fs.readFileSync(SOURCES_FILE, 'utf-8'));
    if (!Array.isArray(data.sources)) data.sources = [];
    return data;
  } catch {
    return { version: 1, sources: [] };
  }
}

function save(data: Registry): void {
  fs.mkdirSync(path.dirname(SOURCES_FILE), { recursive: true });
  fs.writeFileSync(SOURCES_FILE, JSON.stringify(data, null, 2) + '\n', 'utf-8');
}

function countFiles(dir: string): number {
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) return 0;
  let count = 0;
  const walk = (d: string) => {
    walkEntries(d, (full, name, stat) => {
      if (name.startsWith('.')) return;
      if (stat.isDirectory()) walk(full);
      else if (stat.isFile()) count += 1;
    });
  };
  try {
    walk(dir);
  } catch {
    return count;
  }
  return count;
}

export function listSources(): SourceEntryWithStatus[] {
  const { sources } = load();
  return sources.map((entry) => {
    const exists = fs.existsSync(entry.path) && fs.statSync(entry.path).isDirectory();
    return { ...entry, exists, file_count: countFiles(entry.path) };
  });
}

export function addSource(rawPath: string, label?: string | null): SourceEntry {
  if (!rawPath || !rawPath.trim()) throw new SourceError('Path is required');
  const resolved = path.resolve(rawPath.trim());
  if (!fs.existsSync(resolved) || !fs.statSync(resolved).isDirectory()) {
    throw new SourceError(`Not a directory: ${resolved}`);
  }

  const rawResolved = path.resolve(RAW_DIR);
  if (resolved === rawResolved || resolved.startsWith(rawResolved + path.sep)) {
    throw new SourceError('Cannot add data/raw/ (or a folder inside it) as a source');
  }
  if (rawResolved.startsWith(resolved + path.sep)) {
    throw new SourceError('Cannot add a folder that contains data/raw/ as a source');
  }

  const data = load();
  if (data.sources.some((entry) => path.resolve(entry.path) === resolved)) {
    throw new SourceError('This folder is already registered');
  }

  const existingNames = new Set(data.sources.map((entry) => entry.link_name));
  const linkName = uniqueLinkName(label || path.basename(resolved), existingNames);

  const entry: SourceEntry = {
    id: crypto.randomBytes(6).toString('hex'),
    label: (label || path.basename(resolved)).trim() || path.basename(resolved),
    path: resolved,
    link_name: linkName,
    enabled: true,
    added_at: new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
  };
  data.sources.push(entry);
  save(data);
  syncSymlinks();
  return entry;
}

export function removeSource(id: string): boolean {
  const data = load();
  const removed = data.sources.find((entry) => entry.id === id);
  if (!removed) return false;
  data.sources = data.sources.filter((entry) => entry.id !== id);
  save(data);
  removeMirror(path.join(RAW_DIR, removed.link_name));
  syncSymlinks();
  return true;
}

export function setEnabled(id: string, enabled: boolean): SourceEntry | null {
  const data = load();
  const entry = data.sources.find((item) => item.id === id);
  if (!entry) return null;
  entry.enabled = enabled;
  save(data);
  syncSymlinks();
  return entry;
}

function removeMirror(root: string): void {
  let stat;
  try {
    stat = fs.lstatSync(root);
  } catch {
    return;
  }
  if (stat.isSymbolicLink()) {
    fs.unlinkSync(root);
    return;
  }
  if (!stat.isDirectory()) return;

  const entries: string[] = [];
  const walk = (d: string) => {
    for (const child of fs.readdirSync(d, { withFileTypes: true })) {
      const full = path.join(d, child.name);
      entries.push(full);
      if (child.isDirectory() && !child.isSymbolicLink()) walk(full);
    }
  };
  walk(root);
  // Deepest first so a directory is only rmdir'd once it's empty.
  entries.sort((a, b) => b.split(path.sep).length - a.split(path.sep).length);
  for (const entry of entries) {
    const entryStat = fs.lstatSync(entry);
    if (entryStat.isSymbolicLink()) {
      fs.unlinkSync(entry);
    } else if (entryStat.isDirectory()) {
      try {
        fs.rmdirSync(entry);
      } catch {
        /* not empty -- leave real content alone */
      }
    }
  }
  try {
    fs.rmdirSync(root);
  } catch {
    /* not empty */
  }
}

function mirrorSource(linkRoot: string, target: string): void {
  fs.mkdirSync(linkRoot, { recursive: true });
  const liveRelPaths = new Set<string>();

  const walkTarget = (dir: string) => {
    walkEntries(dir, (full, _name, stat) => {
      if (stat.isDirectory()) {
        walkTarget(full);
      } else if (stat.isFile()) {
        const rel = path.relative(target, full);
        liveRelPaths.add(rel);
        const dest = path.join(linkRoot, rel);
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        let shouldLink = true;
        try {
          const destStat = fs.lstatSync(dest);
          if (destStat.isSymbolicLink()) {
            const resolvedDest = fs.realpathSync(dest);
            if (resolvedDest === fs.realpathSync(full)) {
              shouldLink = false;
            } else {
              fs.unlinkSync(dest);
            }
          } else {
            shouldLink = false; // a real file already occupies this path
          }
        } catch {
          /* dest doesn't exist -- create it */
        }
        if (shouldLink) {
          try {
            fs.symlinkSync(full, dest);
          } catch {
            /* ignore */
          }
        }
      }
    });
  };
  walkTarget(target);

  // Remove mirrored symlinks whose source file no longer exists.
  const pruneStale = (dir: string) => {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, entry.name);
      const rel = path.relative(linkRoot, full);
      if (entry.isSymbolicLink()) {
        if (!liveRelPaths.has(rel)) fs.unlinkSync(full);
      } else if (entry.isDirectory()) {
        pruneStale(full);
        try {
          fs.rmdirSync(full);
        } catch {
          /* not empty */
        }
      }
    }
  };
  pruneStale(linkRoot);
}

export function syncSymlinks(): void {
  fs.mkdirSync(RAW_DIR, { recursive: true });
  const { sources } = load();
  const wanted = new Map(sources.filter((s) => s.enabled).map((s) => [s.link_name, s.path]));
  const allNames = new Set(sources.map((s) => s.link_name));

  for (const entry of fs.readdirSync(RAW_DIR, { withFileTypes: true })) {
    if (allNames.has(entry.name) && !wanted.has(entry.name)) {
      removeMirror(path.join(RAW_DIR, entry.name));
    }
  }

  for (const [name, target] of wanted) {
    const linkRoot = path.join(RAW_DIR, name);
    try {
      const stat = fs.lstatSync(linkRoot);
      if (stat.isSymbolicLink()) {
        fs.unlinkSync(linkRoot); // replace a stale top-level symlink with a real mirror dir
      } else if (!stat.isDirectory()) {
        continue; // a real file already occupies this name
      }
    } catch {
      /* doesn't exist yet */
    }
    if (!fs.existsSync(target) || !fs.statSync(target).isDirectory()) continue;
    mirrorSource(linkRoot, target);
  }
}
